### PR TITLE
Update tbb

### LINKED ISSUE #5914: Warnings in bundled TBB
with
```
#        CMAKE_CXX_COMPILER:     Clang 5.0.1 on platform Linux x86_64
```
I get the following warning:
```
In file included from /ssd/deal-git/bundled/tbb41_20130401oss/src/tbb/itt_notify.cpp:55:
/ssd/deal-git/bundled/tbb41_20130401oss/src/tbb/tools_api/ittnotify_static.c:283:20: warning: passing an object that undergoes default argument promotion to 'va_start' has undefined behavior [-Wvarargs]
    va_start(args, code);
                   ^
/ssd/deal-git/bundled/tbb41_20130401oss/src/tbb/tools_api/ittnotify_static.c:280:49: note: parameter of type '__itt_error_code' is declared here
static void __itt_report_error(__itt_error_code code, ...)
                                                ^
```
Thoughts?

### PR DESCRIPTION
**Pull request by drwells**
Fixes #5914. This was actually not too hard to do. I think that I fixed all of the warnings (unless, of course, one adds more warnings through `DEAL_II_CXX_FLAGS_DEBUG`)

Regardless of whether or not we update it, it doesn't look like we currently apply one of the required GCC fixes for disabling a dead store optimization: we need to do that either way.

### CODE REVIEW

**Comment by drwells** [comment]
Also of note from #5914: this version is under the Apache license, which I believe is more permissive than the GPL with a linking exception. We need to double-check that our licenses are compliant.

**Comment by Rombur** [comment]
Yes Apache 2.0 is more permissive than what they used before so there shouldn't be any problem.

**Comment by Rombur** [comment]
Apparently Apache 2.0 is compatible with GPL 3.0 but **NOT** GPL 2.0. So some people will have to be careful when they link everything together.

File: bundled/tbb-2018_U2/.gitattributes [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# Set the default behavior, in case people don't have core.autocrlf set.
```
Comment (tamiko) [inline bundled/tbb-2018_U2/.gitattributes]: Can you remove this file, please :-)

**Comment by drwells** [comment]
@tamiko I fixed both; the new version does not have a COPYING file.

Does this license change imply that we need to link things separately?

**Comment by drwells** [comment]
This legal situation is not clear: see, e.g.,

https://opensource.stackexchange.com/questions/1357/can-i-link-a-apache-2-0-library-into-software-under-gplv2

At least we are not mixing code in the same subroutine, but they do end up in the same shared object file.

**Comment by drwells** [comment]
I spent some more time reading up on this and I am no longer convinced that we should do this.

Suppose, for the sake of argument, that we want to create a derived work from TBB. With that derived work we wish to bundle an LGPL-licensed library (deal.II). Suppose that this creates a new derived work (TBB + deal.II). According to

https://tldrlegal.com/license/gnu-lesser-general-public-license-v3-(lgpl-3)

we must distribute the resulting product under the LGPL, which is not possible since it has conflicts with the Apache license.

I think bundling the current version of TBB is okay since it is GPL: users must agree, if they want to use TBB, that they must treat the resulting product as a GPL-licensed library.

My main conclusion is that I have not yet been able to find a clear answer online. I suggest that we hire a lawyer unless someone is able to find a clear answer here.

**Comment by jppelteret** [comment]
Would be of any use trying to find the last release before they changed their license, and seeing if the issue #5914 is corrected there? Then we could at least safely update to that version.

**Comment by drwells** [comment]
We could also try something like

https://fossa.io/pricing/

We would not have to run this more than perhaps once ever, but it would tell us the right answer.

**Comment by Rombur** [comment]
@drwells I don't understand this part 
> we must distribute the resulting product under the LGPL, which is not possible since it has conflicts with the Apache license

My understanding is that there is only a problem with apache + gpl v2. So deal + TBB is fine, there is only a problem if someone uses deal + TBB  + gpl v2 only library.

**Comment by tamiko** [comment]
> there is only a problem if someone uses deal + TBB + gpl v2 only library.

Let's be very careful with the wording here - there will be a problem for someone who wants to <b>distribute</b> deal.II in binary form with bundled tbb. But for distributing a binary copy of deal.II there are already a bazillion of license incompatibilities in place :-/

I don't see a problem with bundling tbb with the new license because internal tbb isn't used in an binary distribution (or we can easily ensure to, e.g., compile and ship an external tbb library).

**Comment by drwells** [comment]
@tamiko I agree that the main issue is binary distribution. To be clear: is it your understanding that we should be OK if all binary distributions from here on do *not* use the bundled TBB?

**Comment by bangerth** [comment]
What should we do with this? @tamiko -- do you agree with @drwells 's last statement?

**Comment by tamiko** [comment]
@drwells @bangerth Yes, I agree with this summary:
 - We are fine to distribute our source tarball as is. [We clearly state what parts of the sources are covered by what library, all licenses allow distribution in source form alongside each other.]
 - An end user (which is 99% of the use cases) has no problem either - they typically neither distribute their sources, nor binaries.
 - For our Mac bundles and VM images: We have to make sure to install an external copy of tbb.
 - In general, every binary distribution is best advised to simply `rm -r bundled` and install external dependencies seperately.

**Comment by bangerth** [comment]
On 03/21/2018 08:55 AM, Matthias Maier wrote:
> # For our Mac bundles and VM images: We have to make sure to install an 
> external copy of tbb.

Or make sure the VMs have the source files in them -- which is what I 
usually do (for many reasons -- for example so that debugging works).

**Comment by tamiko** [comment]
@bangerth 
> Or make sure the VMs have the source files in them -- which is what I usually do (for many reasons -- for example so that debugging works).

That doesn't help. We are not allowed to **distribute a binary library** that links together our lgpl-2.1 covered code and tbb sources.

**Comment by Rombur** [comment]
So can we merge this?

**Comment by bangerth** [comment]
I'm for it!

**Pull request merged**